File: ios/PardusLockApp/UI/Admin/AdminViewModel.swift
import Foundation

enum AdminState {
    case idle
    case loading
    case error(String)
}

@MainActor
final class AdminViewModel: ObservableObject {
    @Published var users: [User] = []
    @Published var state: AdminState = .idle
    @Published var actionError: String? = nil

    let currentUser: User

    init(currentUser: User) {
        self.currentUser = currentUser
    }

    var isLoading: Bool {
        if case .loading = state { return true }
        return false
    }

    // MARK: - Load

    func loadUsers() async {
        state = .loading
        do {
            let response = try await APIClient.shared.getUsers()
            if response.status == "ok" {
                users = response.users ?? []
                state = .idle
            } else {
                state = .error(response.message ?? "Kullanıcılar yüklenemedi.")
            }
        } catch {
            state = .error(error.localizedDescription)
        }
    }

    // MARK: - Add User

    func addUser(username: String, fullName: String, password: String, role: String) async -> Bool {
        do {
            let response = try await APIClient.shared.addUser(
                username: username,
                fullName: fullName,
                password: password,
                role: role
            )
            if response.status == "ok" {
                await loadUsers()
                return true
            } else {
                actionError = response.message ?? "Kullanıcı eklenemedi."
                return false
            }
        } catch {
            actionError = error.localizedDescription
            return false
        }
    }

    // MARK: - Delete User

    func deleteUser(id: Int) async {
        guard id != currentUser.id else {
            actionError = "Kendi hesabınızı silemezsiniz."
            return
        }
        do {
            let response = try await APIClient.shared.deleteUser(id: id)
            if response.status == "ok" {
                users.removeAll { $0.id == id }
            } else {
                actionError = response.message ?? "Kullanıcı silinemedi."
            }
        } catch {
            actionError = error.localizedDescription
        }
    }

    // MARK: - Change Password

    func changePassword(userId: Int, newPassword: String) async -> Bool {
        let trimmed = newPassword.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else {
            actionError = "Şifre boş olamaz."
            return false
        }
        do {
            let response = try await APIClient.shared.changePassword(userId: userId, newPassword: trimmed)
            if response.status == "ok" {
                return true
            } else {
                actionError = response.message ?? "Şifre değiştirilemedi."
                return false
            }
        } catch {
            actionError = error.localizedDescription
            return false
        }
    }
}

File: android/app/src/main/java/com/pardus/lockapp/ui/admin/AdminViewModel.kt
package com.pardus.lockapp.ui.admin

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.pardus.lockapp.data.api.ApiClient
import com.pardus.lockapp.data.model.AddUserRequest
import com.pardus.lockapp.data.model.ChangePasswordRequest
import com.pardus.lockapp.data.model.User
import kotlinx.coroutines.launch

sealed class AdminState {
    object Idle : AdminState()
    object Loading : AdminState()
    data class UsersLoaded(val users: List<User>) : AdminState()
    data class ActionSuccess(val message: String) : AdminState()
    data class Error(val message: String) : AdminState()
}

class AdminViewModel : ViewModel() {

    private val _state = MutableLiveData<AdminState>(AdminState.Idle)
    val state: LiveData<AdminState> = _state

    fun loadUsers(serverUrl: String) {
        _state.value = AdminState.Loading
        viewModelScope.launch {
            try {
                val response = ApiClient.getService(serverUrl).getUsers()
                if (response.isSuccessful && response.body() != null) {
                    _state.value = AdminState.UsersLoaded(response.body()!!.users)
                } else {
                    _state.value = AdminState.Error(
                        ApiClient.parseErrorMessage(response, "Kullanıcılar yüklenemedi")
                    )
                }
            } catch (e: Exception) {
                _state.value = AdminState.Error("Bağlantı hatası: ${e.message}")
            }
        }
    }

    fun addUser(serverUrl: String, username: String, password: String, fullName: String, role: String) {
        _state.value = AdminState.Loading
        viewModelScope.launch {
            try {
                val response = ApiClient.getService(serverUrl)
                    .addUser(AddUserRequest(username, password, fullName, role))
                if (response.isSuccessful && response.body()?.status == "ok") {
                    _state.value = AdminState.ActionSuccess("Kullanıcı eklendi")
                } else {
                    _state.value = AdminState.Error(
                        ApiClient.parseErrorMessage(response, "Kullanıcı eklenemedi")
                    )
                }
            } catch (e: Exception) {
                _state.value = AdminState.Error("Bağlantı hatası: ${e.message}")
            }
        }
    }

    fun deleteUser(serverUrl: String, userId: Int) {
        _state.value = AdminState.Loading
        viewModelScope.launch {
            try {
                val response = ApiClient.getService(serverUrl).deleteUser(userId)
                if (response.isSuccessful && response.body()?.status == "ok") {
                    _state.value = AdminState.ActionSuccess("Kullanıcı silindi")
                } else {
                    _state.value = AdminState.Error(
                        ApiClient.parseErrorMessage(response, "Kullanıcı silinemedi")
                    )
                }
            } catch (e: Exception) {
                _state.value = AdminState.Error("Bağlantı hatası: ${e.message}")
            }
        }
    }

    fun changePassword(serverUrl: String, userId: Int, newPassword: String) {
        _state.value = AdminState.Loading
        viewModelScope.launch {
            try {
                val response = ApiClient.getService(serverUrl)
                    .changePassword(ChangePasswordRequest(userId, newPassword))
                if (response.isSuccessful && response.body()?.status == "ok") {
                    _state.value = AdminState.ActionSuccess("Şifre değiştirildi")
                } else {
                    _state.value = AdminState.Error(
                        ApiClient.parseErrorMessage(response, "Şifre değiştirilemedi")
                    )
                }
            } catch (e: Exception) {
                _state.value = AdminState.Error("Bağlantı hatası: ${e.message}")
            }
        }
    }

    fun resetState() {
        _state.value = AdminState.Idle
    }
}
